src/skins/vector/views/organisms/UserSettings.js

```diff
@@ -15,15 +15,144 @@ limitations under the License.
+            function(resps) {
+                self.setState({
+                    avatarUrl: resps[0].avatar_url,
+                    displayName: resps[0].displayname,
+                    threepids: resps[1].threepids,
+                    phase: self.Phases.Display,
+                });
+
+                // keep a copy of the original state in order to track changes
+                self.setState({
+                    originalState: self.state
+                });
+            },
+            function(error) {
+                var ErrorDialog = sdk.getComponent("organisms.ErrorDialog");
+                Modal.createDialog(ErrorDialog, {
+                    title: "Can't load user settings",
+                    description: error.toString()
+                });
+            }
+        );
+    },
+
+    componentDidMount: function() {
+        this.dispatcherRef = dis.register(this.onAction);
+    },
+
+    componentWillUnmount: function() {
+        dis.unregister(this.dispatcherRef);
+    },
+
+    onSaveClicked: function(ev) {
+        var self = this;
+        var savePromises = [];
+
+        // XXX: this is managed in ChangeAvatar.js, although could be moved out here in order
+        // to allow for the change to be staged alongside the rest of the form.
+        //
+        // if (this.state.originalState.avatarUrl !== this.state.avatarUrl) {
+        //     savePromises.push( UserSettingsStore.saveAvatarUrl(this.state.avatarUrl) );
+        // }
+
+        if (this.state.originalState.displayName !== this.state.displayName) {
+            savePromises.push( UserSettingsStore.saveDisplayName(this.state.displayName) );
+        }
+
+        if (this.state.originalState.threepids.length !== this.state.threepids.length ||
+            this.state.originalState.threepids.every(function(element, index) {
+                    return element === this.state.threepids[index];
+            }))
+        {
+            savePromises.push( UserSettingsStore.saveThreePids(this.state.threepids) );
+        }
+
+        // TODO: do the password check
+
+        self.setState({
+            phase: self.Phases.Saving,
+        });
+
+        q.all(savePromises).then(
+            function(resps) {
+                self.setState({
+                    phase: self.Phases.Display,
+                });
+                self.onClose();
+            },
+            function(error) {
+                self.setState({
+                    phase: self.Phases.Display,
+                });
+                var ErrorDialog = sdk.getComponent("organisms.ErrorDialog");
+                Modal.createDialog(ErrorDialog, {
+                    title: "Can't save user settings",
+                    description: error.toString()
+                });
+            }
+        );        
+    },
+
+    onClose: function(ev) {
+        // XXX: use browser history instead to find the previous room?
+        if (this.props.roomId) {
+            dis.dispatch({
```

Comment:
This limits the reusability of this component if people ever want to do anything else after saving user settings. Might it be worth calling a `prop` instead and delegating the responsibility to the parent component (`RoomView` or `MatrixChat`)?